src/main/scala/piconot/external/semantics/Interpreter.scala

```diff
@@ -0,0 +1,41 @@
+package piconot.external.PicoDSL.interpreter
+
+import piconot.external.PicoDSL.ir._
+import picolib.semantics._
+
+
+package object semantics {
+  var numStates = 0                                          // keeps track of how many states have already been used
```

Comment:
I think Scala usually prefers immutable but I like your use of var here. We had a similar issue with keeping track of the number of states in our code and we decided to go the immutable route. It ended up in a lot more verbose code.